opentracing-proxy/opentracing-proxy/src/main/java/org/zalando/opentracing/proxy/plugin/LogCorrelation.java

```diff
@@ -105,6 +115,19 @@ public void onClosing(final Scope scope, final Span span) {
         baggage.forEach((ignored, key) -> MDC.remove(key));
     }
 
+    @Override
+    public void onClosed(final Scope scope, final Span span) {
```

Comment:
If we go with the above suggestion, then this could be simplified to:

```java
    @Override
    public void onClosed(final Scope scope, final Span span) {
        Optional.ofNullable(contexts.get().poll())
                .ifPresent(previous -> previous.forEach(MDC::put));
    }
```

Reply:
Actually, why wouldn't we do this within `onClosing` directly?

Reply:
Removing MDC keys in `onClosing` and restoring the original ones in `onClosed` leaves a window in between those two methods calls where log lines would still end up without any context.

In fact, the 100% safe solution would be to change the removal within `onClosing` to conditionally restore. That way there is no gap.

Reply:
Something like:

```java
    @Override
    public void onClosing(final Scope scope, final Span span) {
        final Map<String, String> previousContext = restoreContext();

        contextKeys().forEach(key -> {
            @Nullable final String value = previousContext.get(key);
            if (value == null) {
                MDC.remove(key);
            } else {
                MDC.put(key, value);
            }
        });
    }

    private void preserveContext() {
        final Map<String, String> context =
                newHashMapWithExpectedSize(seeds.size() + baggage.size());

        contextKeys()
                .forEach(contextKey -> {
                    @Nullable final String value = MDC.get(contextKey);
                    if (value == null) return;
                    context.put(contextKey, value);
                });

        contexts.get().push(context);
    }

    private Map<String, String> restoreContext() {
        return Optional.ofNullable(contexts.get().poll()).orElse(emptyMap());
    }

    private Stream<String> contextKeys() {
        return Stream.concat(
                seeds.values().stream(),
                baggage.values().stream());
    }
```

Reply:
We can, just added here because `onClosed` was getting called after the scope has been formally closed.
Just to be sure.